etl/steps/data/garden/demography/2023-03-31/population/__init__.py

```diff
@@ -0,0 +1,289 @@
+    log.info("population: adding regions...")
+    regions = [
+        "Europe",
+        "Asia",
+        "North America",
+        "South America",
+        "Africa",
+        "Oceania",
+        "High-income countries",
+        "Low-income countries",
+        "Lower-middle-income countries",
+        "Upper-middle-income countries",
+        "European Union (27)",
+    ]
+    # make sure to exclude regions if already present
+    df = df.loc[~df["country"].isin(regions)]
+
+    # keep sources per countries, remove from df
+    # remove from df: otherwsie geo.add_region_aggregates will add this column too
+    sources = df[["country", "year", "source"]].copy()
+    df = df.drop(columns=["source"])
+
+    # re-estimate region aggregates
+    for region in regions:
+        df = geo.add_region_aggregates(df=df, region=region, population=df)
+
+    # add sources back
+    # these are only added to countries, not aggregates
+    df = df.merge(sources, on=["country", "year"], how="left")
+
+    # add sources for region aggregates
+    # this is done by taking the union of all sources for countries in the region
+    for region in regions:
+        members = geo.list_countries_in_region(region)
+        s = df.loc[df["country"].isin(members), "source"].unique()
+        sources_region = sorted(s)
+        df.loc[df["country"] == region, "source"] = "; ".join(sources_region)
+    return df
+
+
+def add_world(df: pd.DataFrame) -> pd.DataFrame:
+    """Add world aggregate.
+
+    We do this by adding the values for all continents.
+
+    Note that for some years, there is data available for 'World'. This is because some of the datasets
+    contain 'World' data but others don't.
+    """
+    log.info("population: adding World...")
+    df_ = deepcopy(df)
+    year_threshold = df_.loc[df_["country"] == "World", "year"].min()
+    assert (
+        year_threshold == 1950  # This is the year that the UN data starts.
+    ), "Year threshold has changed! Check if HYDE or Gapminder (or UN WPP) have data for 'World' before 1950!"
+    continents = [
+        "Europe",
+        "Asia",
+        "North America",
+        "South America",
+        "Africa",
+        "Oceania",
+    ]
+    # Estimate "World" population for years before `year_threshold` and add to original data.
+    df_ = (
+        df_[(df_["country"].isin(continents)) & (df_["year"] < year_threshold)]
+        .groupby("year", as_index=False)["population"]
+        .sum(numeric_only=True)
+        .assign(country="World")
+    )
+    df = pd.concat([df, df_], ignore_index=True).sort_values(["country", "year"])
+
+    # add sources for world
+    df.loc[df["country"] == "World", "source"] = "; ".join(sorted(SOURCES_NAMES.values()))
+    return df
+
+
+def add_historical_regions(df: pd.DataFrame) -> pd.DataFrame:
+    """Add historical regions.
+
+    Systema Globalis from Gapminder contains historical regions. We add them to the data. These include
+    Yugoslavia, USSR, etc.
```

Comment:
Just to confirm: A country like Russia has data that is extended backwards even in USSR times, right? Otherwise historical regions would need to be added prior to creating region aggregates.

Reply:
Yes, countries like Russia ("current states") extend back in time. Former countries do not replace any of these.